extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java

```diff
@@ -0,0 +1,215 @@
+
+    @SuppressWarnings("unchecked")
+    @NonNullByDefault({}) // TODO Until the ModuleTypeProvider interface is fixed
+    @Override
+    public ModuleType getModuleType(String UID, Locale locale) {
+        if (PublishActionHandler.MODULE_TYPE_ID.equals(UID)) {
+            return getActionType(locale);
+        } else if (PublishedMessageTriggerHandler.MODULE_TYPE_ID.equals(UID)) {
+            return getTriggerType(locale);
+        } else {
+            return null;
+        }
+    }
+
+    @NonNullByDefault({}) // TODO Until the ModuleTypeProvider interface is fixed
+    @Override
+    public Collection<ModuleType> getModuleTypes(Locale locale) {
+        cachedLocale = locale;
+        return Stream.of(getActionType(locale), getTriggerType(locale)).collect(Collectors.toList());
+    }
+
+    private ModuleType getActionType(Locale locale) {
+        final ActionType actionType = new ActionType(PublishActionHandler.MODULE_TYPE_ID, getConfigActionDesc(locale), //
+                "Publish MQTT message", "Publishes a message to an MQTT topic", //
+                null, Visibility.VISIBLE, getInputActionDesc(locale), getOutputTriggerDesc(locale));
+        this.actionType = actionType;
+        return actionType;
+    }
+
+    private ModuleType getTriggerType(Locale locale) {
+        return new TriggerType(PublishedMessageTriggerHandler.MODULE_TYPE_ID, getConfigTriggerDesc(locale),
+                "MQTT message published", "Triggers whenever the given MQTT topic has a new value", //
+                null, Visibility.VISIBLE, getOutputTriggerDesc(locale));
+    }
+
+    private @Nullable List<Output> getOutputTriggerDesc(Locale locale) {
+        Output topicName = new Output(MQTTModuleConstants.INOUT_TOPIC_NAME, MQTTModuleConstants.TYPE_TOPIC, //
+                "Topic", "A topic that was triggered from", null, null, null);
+        Output topicValue = new Output(MQTTModuleConstants.INOUT_TOPIC_VALUE, MQTTModuleConstants.TYPE_STRING, //
+                "Topic", "The received topic value if triggered from a topic", null, null, null);
+        Output brokerConnection = new Output(MQTTModuleConstants.INOUT_BROKER_CONNECTION,
+                MQTTModuleConstants.INOUT_BROKER_CONNECTION, //
+                "Broker connection", "The broker connection", null, null, null);
+        return Stream.of(topicName, topicValue, brokerConnection).collect(Collectors.toList());
+    }
+
+    private @Nullable List<Input> getInputActionDesc(Locale locale) {
+        Input topicName = new Input(MQTTModuleConstants.INOUT_TOPIC_NAME, MQTTModuleConstants.TYPE_TOPIC, //
+                "Topic", "A topic that was triggered from", null, false, null, null);
+        Input topicValue = new Input(MQTTModuleConstants.INOUT_TOPIC_VALUE, MQTTModuleConstants.TYPE_STRING, //
+                "Topic", "The received topic value if triggered from a topic", null, false, null, null);
+        Input brokerConnection = new Input(MQTTModuleConstants.INOUT_BROKER_CONNECTION,
```

Comment:
The broker is a configuration parameter of the action, it must not be additionally be an input.

Reply:
If a rule got triggered by the MQTT trigger, that exact broker connection can be used for the MQTT action and I see this as valid use case.
A boolean configuration "Trigger connection overrides configured broker" is missing to make this explicit. Wdyt?